From 2f181bbd819d88dc2ff43486deaa6eb6180634c1 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Wed, 7 Aug 2024 10:37:11 -0700
Subject: [PATCH] hostapd: enable HE when EHT is set for channel switch

Channel switch operation fails if 'eht' option is given without 'he'
as hostapd_config_check() returns following error:
"Cannot set ieee80211be without ieee80211ax"

The issue is seen only when switching within/to DFS channels because
hostapd_switch_channel_fallback() does not set 'ieee80211ax' in
hostapd configuration unless 'he_enabled' is set for the new channel.

Set 'he_enabled' in struct hostapd_freq_params for channel switch
when EHT is enabled.

Patch-dependency: None

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 src/ap/ctrl_iface_ap.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index 698ace8..3acce24 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -1403,8 +1403,9 @@ int hostapd_parse_csa_settings(const char *pos,
	settings->freq_params.punct_bitmap = settings->punct_bitmap;
 	settings->freq_params.ht_enabled = !!os_strstr(pos, " ht");
 	settings->freq_params.vht_enabled = !!os_strstr(pos, " vht");
-	settings->freq_params.he_enabled = !!os_strstr(pos, " he");
 	settings->freq_params.eht_enabled = !!os_strstr(pos, " eht");
+	settings->freq_params.he_enabled = !!os_strstr(pos, " he") ||
+					   settings->freq_params.eht_enabled;
 	settings->block_tx = !!os_strstr(pos, " blocktx");
	settings->handle_dfs = !!os_strstr(pos, " handle_dfs");
 #undef SET_CSA_SETTING
